src/devices/src/virtio/mmio.rs

```diff
@@ -343,542 +343,575 @@ impl BusDevice for MmioTransport {
+            LittleEndian::write_u32(&mut buf[..], 16);
+            d.write(0x38, &buf[..]);
+            LittleEndian::write_u32(&mut buf[..], 1);
+            d.write(0x44, &buf[..]);
+        }
+        assert!(d.are_queues_valid());
+        assert!(!d.locked_device().is_activated());
+
+        // Device should be ready for activation now.
+        set_device_status(
+            d,
+            device_status::ACKNOWLEDGE
+                | device_status::DRIVER
+                | device_status::FEATURES_OK
+                | device_status::DRIVER_OK,
+        );
+        assert_eq!(
+            d.device_status,
+            device_status::ACKNOWLEDGE
+                | device_status::DRIVER
+                | device_status::FEATURES_OK
+                | device_status::DRIVER_OK
+        );
+        assert!(d.locked_device().is_activated());
+    }
+
+    #[test]
+    fn test_bus_device_reset() {
+        let m = GuestMemory::new(&[(GuestAddress(0), 0x1000)]).unwrap();
+        let mut d = MmioTransport::new(m, Arc::new(Mutex::new(DummyDevice::new()))).unwrap();
+        let mut buf = vec![0; 4];
+
+        assert!(!d.are_queues_valid());
+        assert!(!d.locked_device().is_activated());
+        assert_eq!(d.device_status, 0);
+        activate_device(&mut d);
+
+        // Marking device as FAILED should not affect device_activated state
+        LittleEndian::write_u32(&mut buf[..], 0x8f);
+        d.write(0x70, &buf[..]);
+        assert_eq!(d.device_status, 0x8f);
+        assert!(d.locked_device().is_activated());
+
+        // Nothing happens when backend driver doesn't support reset
+        LittleEndian::write_u32(&mut buf[..], 0x0);
+        d.write(0x70, &buf[..]);
+        assert_eq!(d.device_status, 0x8f);
+        assert!(d.locked_device().is_activated());
+    }
+
+    #[test]
+    fn test_get_avail_features() {
+        let dummy_dev = DummyDevice::new();
+        assert_eq!(dummy_dev.avail_features(), dummy_dev.avail_features);
+    }
+
+    #[test]
+    fn test_get_acked_features() {
+        let dummy_dev = DummyDevice::new();
+        assert_eq!(dummy_dev.acked_features(), dummy_dev.acked_features);
+    }
+
+    #[test]
+    fn test_set_acked_features() {
+        let mut dummy_dev = DummyDevice::new();
+
+        assert_eq!(dummy_dev.acked_features(), 0);
+        dummy_dev.set_acked_features(16);
+        assert_eq!(dummy_dev.acked_features(), dummy_dev.acked_features);
+    }
+
+    #[test]
+    fn test_ack_features_by_page() {
+        let mut dummy_dev = DummyDevice::new();
+        dummy_dev.set_acked_features(16);
+        dummy_dev.set_avail_features(8);
+        dummy_dev.ack_features_by_page(0, 8);
+        assert_eq!(dummy_dev.acked_features(), 24);
+    }
+
+    #[test]
+    fn test_set_avail_features() {
+        let mut dummy_dev = DummyDevice::new();
+        assert_eq!(dummy_dev.avail_features(), 0);
+        dummy_dev.set_avail_features(4);
+        assert_eq!(dummy_dev.avail_features(), 4);
+    }
```

Comment:
`set_avail_features()` is a test function. I presume you are adding it to help you with testing stuff, in which case we shouldn't have a separate test for it.